### DISCUSSION TITLE
Oblique Angle Beam Driven Wakefield

### DISCUSSION THREAD

**Discussion by abigailalexjames** [SmileiPIC/Smilei#698] [General]
Hi Smilei team, 

I'm starting this thread as a continuation of the discussion on the chat. 

I'm attempting to implement a beam driven wakefield at an oblique angle to the x-axis and thanks to the advice so far have managed to get this started. I'll copy the information from the chat so far below:

> abigailjames: Hi there. I'm new to Smilei and had a question I hope you could help with. I'm interested in driving a wakefield at an angle. Both laser and beam driven. I can see in the documentation how to initialise the laser at an angle but not sure about the electron bunch. If you could direct me to anything that could help that would be great. I'm working off the tutorial for the Field initialisation for a relativistic electron bunch. 

> beck-llr: Hi, in that tutorial you can see that the initial positions and momenta of the beam particles are initialized via numpy arrays. You can modify these properties manually so that the beam propagates at an angle (just operate a rotation centered at the center of the beam for instance). That implies a little bit of python manipulation from you. Nevertheless, keep in mind that the moving window can only move along the x direction. So your beam (or laser) are bound to exit the simulation domain at some point if they're not propagating along x.

> fredpz: The relativistic field initialisation will cause problems i believe. It assumes a beam along x. Is it possible instead to have your plasma density profile modified with an angle?
This might actually even work with the AM geometry but I'm not sure

> abigailjames: Hi Fred. Thanks for this. I gave it a try today and you are indeed correct. It was not happy about it. The sim still ran and provided outputs but everything is propagating along x. When I check the slurm output there's a lot of 'computing El from scalar potential, relativisitc Poisson problem'. Would I be right in assuming that it's seen the rotations I placed on the position and momentum, thought nuh-uh and went ahead and assumed an x propagation so that I still get an output that looks good just not in the direction I'm looking for? Or have I misunderstood?
As to the plasma modification; I'm not sure. Ultimately what I need is a probe beam propagating through the wakefield at an oblique angle. I could place the probe at an angle but then I have issues with moving window and propagation distance.

> fredpz: It's not really that the solver forces a beam along x, but that it cannot support fields that are not compatible with this. Thus your result. Can you have the probe beam done as post processing?

> abigailjames: I've just had a look at the documentation and not sure if that will work. The overall simulation/experiment uses chirped pulses to impart a frequency shift onto the probe beam due to interaction with the Wakefield. This can then be extracted from an interference pattern to reconstruct the structure. I can revert to using a laser as a driver if necessary but computationally and with regard to lower density the particle driver is more efficient.

> z10f: One possibility to have a beam with an arbitrary direction of propagation is to turn off all poisson solvers (relativistic and classic). This way you will have zero field and the continuity equation will ensure that the initial divergence of E and B is maintained
This means that with zero field initially, the simulation will behave as if there was a completely identical, immobile bunch of positron neutralising the e- beam. With the e-beam moving in any direction the "positrons" (impossible to see in any way) will remain immobile and the electron beam will be surrounded by its self field
This charge separation can create significant deviations on the dynamics of the electron beam, so if your e- beam has a really high energy probably it will not feel them a lot. Low energy beams will have strong deviations
for the moment I don't see other ways of having a beam propagating along an arbitrary direction

> abigailjames: Hi again! I just wanted to swing back to this suggestion and say that it worked. I switched to 2d cartesian and rotated the arrays and I have an electron bunch at an angle. I need to do some work on shaping the bunch but it does drive a wakefield (initially and then the fields go a bit wild).
Technically it also worked in AMcylindrical but you get two bunches above and below the axis.

> z10f: Hello again, indeed the AMcylindrical geometry probably is not suited for a wakefield with an angle. That geometry will limit the mode numbers on the one you chose, so the more the bunch will drift from the axis or lose symmetry, the wilder (and inexact) the results will be
probably the best choice would be 2D cartesian
I would try to use PML to absorb fields on the borders
but I don't know if the wild fields you mention are at the borders

> abigailjames: They're at the back of the wake, the first bubble forms but then the field interactions seem to blow it out. The boundary conditions are PML so I think it's due to the not ideal shape of the bunch. Here's a pic if you're curious. I'll keep working on the beam shaping and adding the lasers but I'm hopeful that this will work out and that I can run it in 3d. The wake doesn't have to be perfect, it's more important that the laser picks up a frequency shift as it passes.
![image](https://github.com/SmileiPIC/Smilei/assets/74933293/d75bdb6c-0b8c-4c3f-b304-e66b3f258c34)

> z10f: That's really really nice
Apart from the asymmetry due to the angle of the driver, I don't see strange phenomena in the density. I mean, I know you have those waves at the borders of the window, but that's because of the finite size of the window and of the plasma
with a larger window and plasma you should see something different

**Comment by Z10Frank** [comment]
Have you tried with a very large window (and plasma) to see if those effects at the border of the plasma are mitigated?
Those regularly spaced lines in the wake may come from the `position_initialization=regular` of the particles. With a `position_initialization=random` and/or more particles per cell they should gradually disappear (although the density will be more noisy with initially random positions).

**Reply by abigailalexjames** [reply]
Hi @Z10Frank 

Sorry for the delay! I was working on the chirp and just came back to the electron bunch. 

I did try with a larger sim window and plasma however the effects are still there. I think it's the shaping of the electron bunch and how it's evolving. To be fair, I simply took the AMcylindrical code and tried to change it to a 2d version so I don't think I adapted the bunch parameters properly. I've been dealing with laser driven up to now so just starting to understand the bunch parameters. 

This is the bunch from the example deck on the website:
![image](https://github.com/SmileiPIC/Smilei/assets/74933293/550859cd-1b94-4fd9-87ad-54edce8c0b3f)

And this is the bunch from my deck:
![image](https://github.com/SmileiPIC/Smilei/assets/74933293/68fc535c-12cd-4e75-878b-1c6f001518b8)

I've had a look at the benchmarks on github and found a 3d electron bunch wake that doesn't use the input from array so I'm working on adapting that one now to see if it works any better.

**Comment by Z10Frank** [comment]
I am a little confused here because your new figure has the same window sized of the first one you have posted, and the bunch is near the lower y border now, probably it is a test without a larger grid. 
Anyway, please feel free to share a namelist for the oblique beam driven part in 2D cartesian, in case I can make some tests.

**Reply by abigailalexjames** [reply]
Hi Francesco,

Yes apologies, that was the same grid size as the original deck just to show the beam shape. 
Here's an image of the full wake with a larger grid and plasma and the namelist I used to get this: 
[Beam_Angle_2d_V01.txt](https://github.com/SmileiPIC/Smilei/files/14468576/Beam_Angle_2d_V01.txt)

![Beam_Angle_2d_V01](https://github.com/SmileiPIC/Smilei/assets/74933293/d168a352-827d-41af-83de-d64eae1375c3)

I need to look at the rotation on the momentum arrays for the electron bunch as I think they may need more than the simple rotation.

**Comment by Z10Frank** [comment]
Thank you! So what are the effects that are interfering with your study? The reflections at the lower border of the plasma? Or maybe the modulations at high frequency in the wake?

Also, in this rotated case I would try a `dy=dx` (and adjusting the `dt` to satisfy the CFL condition accordingly). It will not change the border effects, but it may treat the x and y directions more symmetrically.

The rotation seems correct, but I can look more in detail.

**Comment by Z10Frank** [comment]
I confirm that the rotation of the bunch seems correct, I tested some modifications here and there:
- `dy=dx` to be more symmetric at least on the axes x and y. The dt was adapted to be slightly smaller than the CFL limit;
- larger plasma and simulation window;
- set the speed of the moving window to the x component of the bunch;
- changed the vertices of the 2D Probe

Feel free to try the resulting namelist, you may want to increase the `simulation_time`:
[Oblique_PWFA.txt](https://github.com/SmileiPIC/Smilei/files/14483653/Oblique_PWFA.txt)

Also, I suggest to experiment with `position_initialization="random"` for the plasma as in this modified namelist or go back to `"regular"`. In both cases increasing the number of macro-particles per cell should reduce the artefacts due to the low number of macro-particles including the noise.

**Reply by abigailalexjames** [reply]
This is great thanks! It's good to know that it makes sense. Thank you for the notes and recommendations. I ran a test and fixing the dy=dx fixed a problem I had been having with the angle. It's more oblique now which is great. 
I haven't increased the resolution yet as I need to get the lasers into the simulation first. I seem to be struggling with this! I have  test script of chirped pulses in a vacuum which works fine. I thought if I redefined the parameters using the plasma frequency as the reference that it should all work the same but the pulse is not appearing as I expect. 
This is the pulse as I expect: 
![GetImage (1)](https://github.com/SmileiPIC/Smilei/assets/74933293/091e4b9e-4c3b-4577-981d-8732bd39a370)
And this is it once I redefine the reference frequency:
![image](https://github.com/SmileiPIC/Smilei/assets/74933293/aed835b9-71d3-4900-98a0-60cdefd4cf8a)

I've also tried to go the other way and alter the namelist for the wakefield to use the laser frequency as a reference but that also breaks. I've gone through all the inputs and am confident my numbers are right but I'm missing something in the scaling and can't see what it could be. I'll keep working on it but have you come across this before?

**Comment by Z10Frank** [comment]
Thanks for the feedback!
To add a laser pulse of frequency `omega_L` in the namelist for the oblique PWFA (where the reference frequency is `omega0`), I would define a `Laser` (either with a shortcut or with a Python function for the boundary conditions) with normalized frequency `omega=omega_L/omega0`.

**Comment by Z10Frank** [comment]
And of course for a laser propagating in the x direction you need to ensure that your `dx` (and `dt`) can properly sample the laser wavelength (and period).

**Comment by abigailalexjames** [comment]
Hi There:

This simulation is finally coming along, I've been able to add the chirped laser and everything seems to be in place for a plasma density of 10^16 cm^-3 with a lambda_p of 333 micron. 

I'm having an issue with scaling it to other densities. When I try to go to n0 = 10^18cm^-3 with a lambda_p of 33 micron, I can't seem to get the electron bunch scaling correct. 

From the Smilei Chat we talked about the QTilde factor and the n_p/n0 value. My understanding is that I need an np/n0 >1 to drive a wakefield, and the bunch charge can be scaled using QTilde = e*np*sigma_x*sigma_y*sigma_z*((2pi)^1/3)

If I use this equation and the parameters for the bunch in the tutorial example I get an np/n0 value of 2.35. If I then apply this to my new plasma density and my bunch size I get a value of QTilde ~ 2000pC for a sigma_x = 23 micron and sigma_y = 4 micron.  However when I run this I get this output:
![GetImage (3)](https://github.com/SmileiPIC/Smilei/assets/74933293/26639390-a548-4eae-8df2-0aa07d2333f8)


My understanding is that I want my electron bunch to be on the same order as the lamba_p I'm using. However in the example for a lambda_p of 333 micron the sigma_x is 50 micron and sigma_y is 9 micron with a QTilde of 240pC. If I scale the bunch by a factor of 10, the same for a smaller lambda_p (sigma_x = 5 micron, sigma_y = 0.9 micron) and a QTilde of 24pC I get this result: (Hard to see the bunch but it is there and it does drive a Wakefield):
![GetImage (4)](https://github.com/SmileiPIC/Smilei/assets/74933293/eed8104b-c097-4279-a6a5-a081f8a968f5)
![GetImage (5)](https://github.com/SmileiPIC/Smilei/assets/74933293/849ac5fe-ad92-410f-a305-36a0ad71459c)
![GetImage (6)](https://github.com/SmileiPIC/Smilei/assets/74933293/7f9ea6ab-934d-4800-b4e1-cb2abd883dbe)

My main problem right now is that I can't explain why it works for this scale of the bunch and why it doesn't for the other bunch. 

I'm wondering if something is happening with the momentum and if I need to adjust anything in the arrays?

**Reply by Z10Frank** [reply]
Unfortunately there is no guarantee that the scaling will be perfect, but first I would try to study what is happening to the bunch. For example, if you ask for the `Rho_electronbunch` in the `Field` or `Probe`  diagnostic (if the `Species` name is `electronbunch`) you may have some info by sliding over the time steps. Maybe the bunch is completely depleted during the interaction, or diffracts too quickly, this should be visible. 

Also, if you decrease the plasma wavelength the "best" bunch length to drive a wakefield is shrinked. As for LWFA, the longitudinal size of the driver should be of the order of the plasma wavelength. If the bunch is too long, it risks serious self-modulation. I see a bunch of 50 um in the tutorial, for a plasma wavelength of 333 um. For a plasma wavelength of 50 um I would try a similar ratio, e.g. with a bunch of length ~10 um. This would result in a lower charge to maintain the same QTilde I guess. 

To make the bunch more rigid, just for test, you can try increasing its energy, this would slow down its evolution under the effect of propagation in the plasma.

**Reply by abigailalexjames** [reply]
Ok that's great, thank you. I just wanted to make sure I wasn't missing something. If it's a case of playing around with it until it's reasonable that's perfect.

**Comment by abigailalexjames** [comment]
I recently spoke with someone who suggested implementing a gradual acceleration to the electron bunch rather than using the relativistic gamma factor. This should allow for the development of self consistent fields. They also suggested to allow it to run for enough time that the charge region get's left behind and is then removed by the moving window. 
It's not something I need for my simulation but thought I'd mention it, and if I have the time I'll come back to it. It may make the simulation cleaner overall.

**Reply by Z10Frank** [reply]
I agree that allowing the moving window take care of the "virtual" charge would be good, but its effect may be localised. The "virtual" charge left behind the bunch is immobile and its effect attenuates quickly with the distance, I don't know if it is in a zone of interest for your set-up where it can give an interference.

I am not sure of what you mean with gradual acceleration of the bunch: once you set the velocities of the bunch particles, they can only slow down since they are driving a wakefield. There is no self-consistent way to selectively accelerate the bunch particles with a fixed arbitrary acceleration when the simulation is ongoing.
